File: Search_a_2D_Matrix_II.cpp
class Solution {
public:
    /**
     * @param matrix: A list of lists of integers
     * @param target: An integer you want to search in matrix
     * @return: An integer indicate the total occurrence of target in the given matrix
     */
    int searchMatrix(vector<vector<int> > &matrix, int target) {
        // write your code here
        int m = matrix.size();
        if (!m){
            return 0;
        }
        int n = matrix[0].size();
        if (!n){
            return 0;
        }
        int i = 0, j = n - 1;
        int count = 0;
        while ( i < m && j > -1){
            if (matrix[i][j] == target){
                ++count;
                ++i;
                --j;
            }else if (matrix[i][j] < target){
                ++i;
            }else {
                --j;
            }
        }
        return count;
    }
};

File: Search a 2D Matrix II.py
class Solution:
    """
    @param matrix: An list of lists of integers
    @param target: An integer you want to search in matrix
    @return: An integer indicates the total occurrence of target in the given matrix
    """
    def searchMatrix(self, matrix, target):
        # write your code here
        
        if not matrix:
            return 0
        m, n = len(matrix), len(matrix[0])
        left, right = 0, n - 1
        while left + 1 < right:
            mid = left + (right - left) / 2
            if matrix[m - 1][mid] < target:
                left = mid
            else:
                right = mid
        
        row, col = m - 1, right
        if matrix[m - 1][left] >= target:
            col = left
        
        ret = 0
        while row >= 0 and col < n:
            if matrix[row][col] == target:
                ret += 1
                row -= 1
                col += 1
            elif matrix[row][col] > target:
                row -= 1
            else:
                col += 1
        return ret
        